File: Go/Data Structure/Array/staticArray.go
package main

import "fmt"

type Array struct {
	array []int
	len   int
}

const size int = 10

func newArray() *Array {
	return &Array{
		array: make([]int, size),
		len:   0,
	}
}

func (array *Array) Push(item int) {
	if array.len >= size {
		fmt.Println("Array is full")
		return
	}
	array.array[array.len] = item
	array.len++

	return
}

func (array *Array) Delete(index int) []int {
	if index < 0 || index > array.len {
		fmt.Println("Index out of Range")
		return nil
	}
	for i := index; i < len(array.array)-1; i++ {
		array.array[i] = array.array[i+1]
	}
	array.len--
	array.array[array.len] = 0

	return array.array[:array.len]
}

func (array *Array) Traverse(fn func(int) int) []int {
	var resultArray []int

	for i := 0; i < len(array.array); i++ {
		resultArray = append(resultArray, fn(array.array[i]))
	}
	return resultArray
}

func (array *Array) GetLength() int {
	return array.len
}

func double(num int) int {
	return num * num
}

// func main() {
// 	array := newArray();

// 	for i := 1; i <= 10; i++ {
// 		array.Push(i);
// 	}

// 	fmt.Println(array.array);
// 	fmt.Println(array.len);

// 	fmt.Println(array.Traverse(double));

// 	fmt.Println(array.Delete(0));
// 	fmt.Println(array.Delete(1));
// 	fmt.Println(array.Delete(2));
// 	fmt.Println(array.Delete(3));
// 	fmt.Println(array.Delete(4));
// 	fmt.Println(array.Delete(5));
// 	fmt.Println(array.array);

// 	array.Push(100);
// 	array.Push(200);
// 	array.Push(300);
// 	array.Push(400);
// 	array.Push(500);
// 	array.Push(600);

// 	fmt.Println(array.array);
// }

File: Nim/Data Structure/Array/static_array.nim
const size = 10

type
    Array = ref object
        array: array[size, int]
        length: int


proc new_array(): Array = 
    Array(length: 0)

proc push(self: Array, item: int) = 
    if self.length >= size: 
        echo "Array is full"
        return
    self.array[self.length] = item
    self.length += 1
    return

proc delete(self: Array, index: int): seq[int] = 
    if index < 0 or index >= self.length: 
        echo "Index out of range"
        return @[]

    for i in index ..< self.length - 1:
        self.array[i] = self.array[i + 1]
    
    self.length -= 1
    self.array[self.length] = 0

    result = @[]

    for i in 0 ..< self.length:
        result.add self.array[i]

proc get_length(self: Array): int = 
    return self.length

proc traverse(self: Array, fn: proc (x: int): int): seq[int] =
    result = @[]
    for i in 0 ..< self.length:
        result.add fn(self.array[i])
    
proc double(x: int): int = 
    x * x

when isMainModule:
    let arr = new_array()

    for i in 1..10:
        arr.push(i)

    
    echo arr.array

    let new_arrays = arr.traverse(double)

    echo new_arrays